ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/worldstate/PrunerTest.java

```diff
@@ -0,0 +1,165 @@
+    final Hash markBlockStateRootHash = newBlock.getHeader().getStateRoot();
+    verify(markSweepPruner, never()).mark(markBlockStateRootHash);
+    verify(markSweepPruner, never()).sweep();
+
+    options = options.setBlockNumber(2L).setParentHash(newBlock.getHash());
+    newBlock = gen.block(options);
+    receipts = gen.receipts(newBlock);
+    blockchain.appendBlock(newBlock, receipts);
+    verify(markSweepPruner).mark(markBlockStateRootHash);
+    verify(markSweepPruner, never()).sweep();
+
+    options = options.setBlockNumber(3L).setParentHash(newBlock.getHash());
+    newBlock = gen.block(options);
+    receipts = gen.receipts(newBlock);
+    blockchain.appendBlock(newBlock, receipts);
+    verify(markSweepPruner).sweep();
+  }
+
+  @Test
+  public void abortsPruningWhenFullyMarkedBlockNoLongerOnCanonicalChain() {
+    final Block genesisBlock = gen.genesisBlock();
+    final BlockchainStorage blockchainStorage =
+        new KeyValueStoragePrefixedKeyBlockchainStorage(
+            new InMemoryKeyValueStorage(), new MainnetBlockHeaderFunctions());
+    final DefaultMutableBlockchain blockchain =
+        new DefaultMutableBlockchain(genesisBlock, blockchainStorage, metricsSystem);
+
+    // start pruner so it can start handling block added events
+    final Pruner pruner = new Pruner(markSweepPruner, blockchain, mockExecutorService, 0, 1);
+    pruner.start();
+
+    /*
+     Set up pre-marking state:
+      O <---- this is a fork as of now (non-canonical)
+      |  O <- marking of the full state trie will begin at this block
+       \/
+       O <--- the common ancestor when the reorg happens
+    */
+    BlockOptions options =
+        new BlockOptions().setBlockNumber(1L).setParentHash(genesisBlock.getHash());
+    final Block canonicalHead = gen.block(options);
+    List<TransactionReceipt> receipts = gen.receipts(canonicalHead);
+    blockchain.appendBlock(canonicalHead, receipts);
+
+    options = new BlockOptions().setBlockNumber(1L).setParentHash(genesisBlock.getHash());
+    final Block forkBlock = gen.block(options);
+    receipts = gen.receipts(forkBlock);
+    blockchain.appendBlock(forkBlock, receipts);
+    /*
+      Cause reorg:
+      O <---- this block causes a reorg; this branch becomes canonical
+      |
+      O <---- which means that state here is referring to nodes from the common ancestor block,
+      |       which is not fully marked
+      |  O <- because this was block at which marking began
+       \/
+       O
+    */
+    options = new BlockOptions().setBlockNumber(2L).setParentHash(forkBlock.getHash());
+    final Block reorgCausingBlock = gen.block(options);
+    receipts = gen.receipts(reorgCausingBlock);
+    blockchain.appendBlock(reorgCausingBlock, receipts);
+    verify(markSweepPruner).mark(canonicalHead.getHeader().getStateRoot());
+    verify(markSweepPruner, never()).sweep();
+  }
+
+  @Test
+  public void shouldThrowIllegalArgumentExceptionCorrectly() {
```

Comment:
nit: A better name would include something about what's being tested, not just the expected outcome: e.g. `shouldRejectInvalidOptions`